### PR TITLE
feat(React SSO): Support simple OIDC authentication in React

### PR DESCRIPTION
**Pull request by jjoyce0510**
**Scope**
This change impacts on the React application / GQL API served from `datahub-frontend` 

**Changes**
This PR introduces changes required to support OIDC authentication in the DataHub React app. We build on top of [Pac4j](https://github.com/pac4j/pac4j), a popular authentication + security library for Java. This will permit us to extend to support other auth methods like LDAP direct, SAML, among others as we evolve the app. 

I've enabled this flow via the introduction of a new "/authenticate" endpoint in the Play server which the React application delegates to in order to perform authentication. In essence, this endpoint provides server-driven instruction about how to handle the auth flow. In cases where OIDC is _not_ enabled, a simple redirect to the traditional JAAS login experience is performed. In cases where OIDC is enabled, a redirect to the configured Identity Provider is issued if the current request is not authenticated. This initiates an indirect flow wherein the Identity Provider is delegated auth responsibility, and redirects back to DataHub on successful authentication of an end user. DataHub subsequently fetches the user's profile attributes and extracts a DataHub username from this information. Finally, DataHub sets 2 authentication cookies, one visible within the client app and another invisible. 

A major change introduces in this PR is removal of the LogIn mutation from the GQL API exposed within `datahub-frontend`. Now, the entire GQL API is protected by an authentication blanket. This makes is easier for the server to drive the authentication experience based on how it is configured and ensures that the GQL is completely secure. 

For detailed information about configuring OIDC auth, see [react-oidc.md](https://github.com/linkedin/datahub/blob/49eddaa10a35d681a0c4eedf4f01db522e5ee055/docs/authentication/react-oidc.md). 

**Validation**
The OIDC flow has been E2E validated using Okta & Google. 

## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

**Comment by shirshanka** [comment]
@jjoyce0510 : just flagging that there are a few conflicts you need to rebase for.

**Review by gabe-lyons — APPROVED** [review]

File: datahub-web-react/src/app/shared/NoPageFound.tsx [hunk 1]
```diff
@@ -1,6 +1,17 @@
...
+
+import { isLoggedInVar } from '../auth/checkAuthStatus';
+import { PageRoutes } from '../../conf/Global';
 
 export const NoPageFound = () => {
+    const isLoggedIn = useReactiveVar(isLoggedInVar);
+
+    if (!isLoggedIn) {
+        return <Redirect to={PageRoutes.LOG_IN} />;
+    }
```
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply datahub-web-react/src/app/shared/NoPageFound.tsx:13]: Yeah great suggestion!

File: datahub-frontend/run/frontend.env [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # Server Port
-PORT=9001
```
Comment (shirshanka) [inline datahub-frontend/run/frontend.env:2]: intentional?

File: datahub-web-react/src/app/shared/NoPageFound.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply datahub-web-react/src/app/shared/NoPageFound.tsx:12]: Actually this does not achieve what I wanted - I want to ensure that when we are testing locally with yarn run start that /authenticate will end up redirecting to login, as the route is not found.

File: datahub-frontend/run/frontend.env (hunk 2 above)
Comment (jjoyce0510) [reply datahub-frontend/run/frontend.env]: Updated

**Review by shirshanka — APPROVED** [review]
This LGTM! 
Excited to see this landing :)

**Pull request merged**